Keyboard accessibility for the MelodySandbox field

glastonbridge opened the pull request:
This makes the MelodySandbox and gallery easily navigable with the keyboard, including
* key handlers
* focus trap
* shared accessible matrix implementation with the LED matrix field
* styling improvements

Built against the latest master with Blockly keyboard accessibility

https://github.com/user-attachments/assets/c3c9ede7-6ed4-4fde-b6e9-a9dc22bc082f

microbit-robert commented on pxtblocks/fields/field_utils.ts:
```diff
@@ -693,4 +693,123 @@ function inflateJRes(jres: pxt.Map<pxt.JRes | string>): pxt.Map<pxt.JRes> {
+        }
+    }
+
+    // TODO: things to sync
+    // on-off class vs styles
+    // sizes
+    // pointer styles oneditable
```
Has this been addressed or not?

microbit-robert commented:
We've lost the white border around the boxes in the melody editor. I think these are important to keep.

Before: 
<img width="295" alt="Screenshot 2025-05-07 at 11 32 13" src="https://github.com/user-attachments/assets/1539c88d-b2b4-482f-a9f4-f02771f68d56" />

After: 
<img width="300" alt="Screenshot 2025-05-07 at 11 31 49" src="https://github.com/user-attachments/assets/203b656b-6cff-499c-a238-dd1d15f66772" />

microbit-robert commented:
The previous ledmatrix field editor work made the show LEDs block symmetric:
<img width="202" alt="Screenshot 2025-05-07 at 11 34 31" src="https://github.com/user-attachments/assets/7998d42c-d531-4a89-98d9-893d1b5aa973" />

It's now gained extra margin on the right hand side between the LEDs and the block edge:
<img width="187" alt="Screenshot 2025-05-07 at 11 34 13" src="https://github.com/user-attachments/assets/5587d8b0-921a-438f-9069-bb5e8779e220" />

microbit-robert commented:
I get the following error in the console when opening the melody field editor `Error: <rect> attribute ry: A negative value is not valid. ("-5.9")`

microbit-robert commented:
Should we focus an element in the field editor when it opens?

microbit-robert commented on pxtblocks/fields/field_melodySandbox.ts:
```diff
@@ -530,49 +556,167 @@ export class FieldCustomMelody<U extends FieldCustomOptions> extends Blockly.Fie
+            const [x, y] = this.selected;
+            const ctrlCmd = pxt.BrowserUtils.isMac() ? e.metaKey : e.ctrlKey;
+            switch(e.code) {
+                case "KeyW":
```
We no longer need to support the old keyboard navigation keys (WASD) so these can be removed.

microbit-robert commented on pxtlib/melody-editor/melodyGallery.ts:
```diff
@@ -148,32 +155,77 @@ namespace pxtmelody {
 
+            this.previewButtons[i] = rightButton;
+              
+            leftButton.addEventListener("keydown", (e) => {
```
We never remove these event listeners, but it seems to be a common pattern in field_melodySandbox.ts...

microbit-robert commented:
There are linting issues to address. Seems to be mostly trailing spaces.

microbit-robert commented on pxtblocks/fields/field_utils.ts:
```diff
@@ -693,4 +693,123 @@ function inflateJRes(jres: pxt.Map<pxt.JRes | string>): pxt.Map<pxt.JRes> {
+    // sizes
+    // pointer styles oneditable
+    function createCell(x: number, y: number, row: SVGElement) {
+        const tx = scale * x * (cellWidth + cellVerticalMargin) + cellVerticalMargin + padLeft;
+        const ty = scale * y * (cellHeight + cellHorizontalMargin) + cellHorizontalMargin;
```
This is different to the original code. tx uses x and horizontal margins, not vertical. These are mixed up for tx and ty?

microbit-robert replied:
Fixing this fixes the width of the showLEDs block, but breaks the melody field editor layout. This is because these two lines of code are not the same, but have been made the same in the utils function. Note the positions of tx and ty.

melody
const cellG = pxsim.svg.child(this.elt, "g", { transform: `translate(${ty} ${tx})` }) as SVGGElement;

showLEDs
const cellG = pxsim.svg.child(row, "g", { transform: `translate(${tx} ${ty})`, 'role': 'gridcell' });

This needs some careful adjustment, following the changes through.

kmcnaught commented:
Looks good to me. I noted an inconsistency in toggle in #46, but I like this version more.

One request - can we auto-scroll if the grid element is out of view? e.g at 300% zoom here, the bottom couple of rows are out of sight and arrowing down doesn't bring them into view
![melody grid element with bottom rows out of sight](https://github.com/user-attachments/assets/87cb7d60-cdad-4c6e-9e85-6e192a556818)

glastonbridge commented:
Thank you for the reviews! I've addressed the points as best I can.

The melody editor seems to manage lifecycle by constructing everything once when it is opened, and destroying it when it's closed. It's not as subtle as React, but I think trying to remove and add them dynamically without React's infrastructure for managing lifecycle is going to introduce risks in new ways. My instincts also find it kind of haunting though. A good fix would be to migrate to React...

Re: the padding. While I was working on this there was a lot of inconsistency between x/y/horiz/vert/col/row, and I did rewrite some of it because it was the only way to get any clarity while I was developing it. melodySandbox was using vertical measurements in x and horizontal measurements in y, for example: https://github.com/microbit-matt-hillsdon/pxt/blob/489df47afff44756be5d44e056dc39ec835f7a1f/pxtblocks/fields/field_melodySandbox.ts#L530

microbit-robert commented on pxtblocks/fields/field_ledmatrix.ts:
```diff
@@ -437,10 +416,6 @@ export class FieldMatrix extends Blockly.Field implements FieldCustom {
         return `\`\n${FieldMatrix.TAB}${text}\n${FieldMatrix.TAB}\``;
     }
 
-    getFieldDescription(): string {
-        return lf("{0}x{1} LED Grid", this.matrixWidth, this.matrixHeight);
-    }
-
```
Not sure why this has been removed.

microbit-robert commented:
Closing as we have this PR raised in `pxt`. Don't delete branch.
